soh/soh/Enhancements/crowd-control/CrowdControl.cpp

```diff
@@ -144,8 +141,21 @@ void CrowdControl::ListenToServer() {
 
             // If effect is a one off run, let's execute
             if (!incomingEffect->timeRemaining) {
-                EffectResult result =
-                ExecuteEffect(incomingEffect->type.c_str(), incomingEffect->value, false);
+                EffectResult result = EffectResult::Retry;
+                if (incomingEffect->category == "spawn_enemy") {
+                    if (GameInteractor::CanSpawnEnemy()) {
+                        if (GameInteractor::Actions::SpawnEnemyWithOffset(incomingEffect->value[0],
```

Comment:
i think it might be nice to simplify this by having `SpawnEnemyWithOffset` handle the `CanSpawnEnemy` check and return an `EffectResult`

then instead of having nested ifs here, we could just do
```cpp
if (incomingEffect->category == "spawn_enemy") {
    result = GameInteractor::Actions::SpawnEnemyWithOffset(incomingEffect->value[0], incomingEffect->value[1]);
}
```

i'm not sure how/if we're using the `uint8_t` values that `SpawnEnemyWithOffset` returns anywhere else, but switching it to be
```cpp
EffectResult GameInteractor::Actions::SpawnEnemyWithOffset(uint32_t enemyId, int32_t enemyParams) {
    if (!GameInteractor::CanSpawnEnemy()) {
        return EffectResult::Retry;
    }

    Player* player = GET_PLAYER(gPlayState);
    [...]
```
might make sense instead of having all this logic live in `ListenToServer`

Reply:
Going to post my general response here to the notion of having the Apply() and SpawnEnemyWithOffset() functions return the effect result instead.

I've talked this through with @dcvz a fair bit as well, and we agree it's ultimately better to switch over to having one function. It'll make every Apply() function a fair bit bigger to include result declaration and logic, but the more places these effects will be used, the more usage we get out of consolidating them (inside of `GameInteractionEffects`).

One thing is that the debugconsole commands were actually using different criteria for some effects compared to CrowdControl (before this rework), but honestly that's probably something that should be changed anyway, so switching it all over atleast ensures they both function the same.

The idea is that the `GameInteractionEffects` are supposed to always be used in conjunction with the checks, while the `GameInteractor` functions exist to be called by `GameInteractionEffects` and any other code that wants to call these functions while doing whatever checks they need on their own. Consolidation the check into Apply() does further strengthens that intention, so that's also a plus.

tl:dr; will try to move any check inside Apply() and SpawnEnemyWithOffset().